### ISSUE TITLE
Display of matrices in UBO inspector is misleading/incorrect

### ISSUE THREAD

**Issue by cdwfs** [baldurk/renderdoc#800]
As we discussed elsewhere: there are some issues displaying matrix data in the UBO inspector, when compiling HLSL to SPIRV for consumption by Vulkan. Possibly in other contexts as well, but that's what I'm testing at the moment :)

1. GLSL and HLSL have different conventions for matrix type names. In GLSL, `matAxB` = A columns, B rows; in HLSL, `floatAxB` means A rows, B columns. In the "Type" column of the UBO inspector, RenderDoc uses GLSL-style suffixes for matrices, but HLSL style prefixes (`float` vs. `vec`/`mat`). So, an undecorated `float3x4` in HLSL (3 rows, 4 columns) has its Type listed as `float4x3` in RenderDoc (even though its contents are displayed correctly).
2. When expanding a matrix in the UBO inspector, the individual rows of the display are labeled "row0..N", even though by default they're showing the matrix columns. This can cause users to believe that their matrices are transposed, or that non-square matrices have the wrong dimensions.
3. Explicitly decorating a matrix as `row_major` in the HLSL leads to very strange results in the RenderDoc display. For a `row_major float4x4`, the contents are displayed as transposed from their default state, as expected. The "Type" field gives no indication of the non-default matrix order, which might be a nice thing to add. But for a non-square matrix (e.g. `row_major float3x4`) not only are the contents of the matrix swizzled in an odd way, but the offsets of subsequent cbuffer fields seems to be messed up. I haven't gotten to the bottom of exactly what's going wrong here, but something is clearly wrong.

So, some feature requests to help clear this up:

1. For the Type column, RenderDoc should follow one convention or the other consistently. Ideally it could either be inferred from the source language (where possible), or configurable by the user. Or maybe it should use its own convention, orthogonal from either high-level language ("matrix col=C row=R", or something similarly explicit).
2. Label the displayed rows of a matrix as "col0...N" instead of "row0...N". At that point, it may make sense to transpose the matrix visually (so that the columns are actually vertical columns), but I can see cases where I'd want to see the struct in actual memory order, so maybe this too should be under user control? Either way, as long as it's explicit (and accurate) about what it's displaying.
3. Figure out what's up with row_major matrices, because that's just weird.

Thanks as always!

**Comment by baldurk** [comment]
I believe the issues described here have been fixed in the v1.0 release. Feel free to comment here or open a new issue if there's still something broken.

**Comment by cdwfs** [comment]
1. Confirmed as fixed; I see "float4x4 (row major)" for HLSL shaders and "float4x4 (column major)" for GLSL shaders.
2. Confirmed as fixed; the individual matrix fields are still displayed as "row0...N" regardless of matrix ordering, but column-major matrices are displayed transposed vs their in-memory order, so "row" is now the correct label in both cases.
3. Confirmed as... a raging dumpster fire? An undecorated float4x4 in HLSL is shown as "row major" in RenderDoc; a "row_major float4x4" is shown as "column major", and a "column_major float4x4" is displayed in RenderDoc as "row major". And somehow this is the correct, intended behavior (citation: https://github.com/KhronosGroup/glslang/issues/433). I'm going to take it on faith that RenderDoc is doing the Right Thing here, and avert my gaze from this abyss before it consumes my soul.

TL;DR: I am satisfied. Thanks again!
